### PR TITLE
fix(Scripts/Midsummer): Implement Midsummer Bonfire mechanics

### PR DESCRIPTION
**Pull request by sudlud**
- fix visuals for ignite / stamp out bonfire
- spawn campfire spellfocus (and flames trap) when fire is burning
- stamping out bonfire flags player for PVP
- entering a zone with a burning bonfire of own faction provides buff 'Bonfire's Blessing'
-> ignite bonfire immediately provides buff to players within zone
-> stamp out bonfire immediately debuffs players within zone
- ensure 'Bonfire's Blessing' PROC gets deactivated and debuffed after event ends

## Changes Proposed:

This PR proposes changes to:
-  [ ] Core (units, players, creatures, game systems).
-  [x] Scripts (bosses, spell scripts, creature scripts).
-  [x] Database (SAI, creatures, etc).

## Issues Addressed:

- Closes 

## SOURCE:

The changes have been validated through:
- [ ] Live research (checked on live servers, e.g Classic WotLK, Retail, etc.)
- [x] Sniffs (remember to share them with the open source community!)
- [x] Video evidence, knowledge databases or other public sources (e.g forums, Wowhead, etc.)
- [ ] The changes promoted by this pull request come partially or entirely from another project (cherry-pick). **Cherry-picks must be committed using the proper --author tag in order to be accepted, thus crediting the original authors, unless otherwise unable to be found**

https://www.wowhead.com/wotlk/spell=45444/bonfires-blessing#comments

Spell IDs based on sniffs provided by @heyitsbench :)

## Tests Performed:

This PR has been:
- [x] Tested in-game by the author.
- [ ] Tested in-game by other community members/someone else other than the author/has been live on production servers.
- [ ] This pull request requires further testing and may have edge cases to be tested.

## How to Test the Changes:

.event start 1

Alliance Player can stamp out Horde Bonfire
Horde Player can stamp out Alliance Bonfire
any player can reignite Horde / Alliance Bonfire

after first event start, each enemy bonfire provides a quest to stamp out the fire
after finishing the quest, each enemy bonfire offers gossip option to stamp out fire

Visit Alliance Bonfire
Map 0
.go ga 76301
.go ga 76302
.go ga 76303
.go ga 76304
.go ga 76306
.go ga 76307
.go ga 76308
.go ga 76311
.go ga 76312
.go ga 76313
.go ga 76314
.go ga 76316
.go ga 76319
.go ga 76321
Map 1
.go ga 76323
.go ga 76326
.go ga 76329
.go ga 76330
.go ga 76335
.go ga 76339
.go ga 76340
.go ga 76342
.go ga 76343
Map 530
.go ga 76344
.go ga 76345
.go ga 76347
.go ga 76349
.go ga 76350
.go ga 76352
.go ga 76354
.go ga 76356
.go ga 76359
Map 571
.go ga 76362
.go ga 76365
.go ga 76367
.go ga 76368
.go ga 76370
.go ga 76372
.go ga 76375
.go ga 76376

Visit Horde Bonfire
Map 0
.go ga 76300
.go ga 76305
.go ga 76309
.go ga 76310
.go ga 76315
.go ga 76317
.go ga 76318
.go ga 76320
.go ga 76322
Map 1
.go ga 76324
.go ga 76325
.go ga 76327
.go ga 76328
.go ga 76331
.go ga 76332
.go ga 76333
.go ga 76334
.go ga 76336
.go ga 76337
.go ga 76338
.go ga 76341
Map 530
.go ga 76346
.go ga 76348
.go ga 76351
.go ga 76353
.go ga 76355
.go ga 76357
.go ga 76358
.go ga 76360
.go ga 76361
Map 571
.go ga 76363
.go ga 76364
.go ga 76366
.go ga 76369
.go ga 76371
.go ga 76373
.go ga 76374
.go ga 76377

when first visiting a fire it burns and provides the buff 'Bonfire's Blessing' to players with same zone and faction
stamping out bonfire immediately debuffs all players of bonfire's faction within the zone

.additem 23247
Use 'Burning Blossom' near Bonfire to ignite it and immediately buff all players of bonfire's faction within zone

It's NOT necessary to initially visit a zone's bonfire to get the buff (if it's not stamped out)
Also NO preloading maps or Bonfires beeing in range required for any functionality of this PR to work correctly

Use 'Burning Blossom' near any of the Alliance / Horde Bonfires above or near any of the following Ahune / City Bonfires
for additional Fire Festival Bonfire buff

Ahune Bonfire
.go ga 11216
.go ga 11217
these are located within an instance and are initially OFF / not burning
using 'Burning Blossom' provides Fire Festival Bonfire buff and ignites the flames

City Bonfire
.go ga 54943
.go ga 242656
.go ga 242657
.go ga 242658
.go ga 242659
.go ga 242660
.go ga 242661
.go ga 242662
.go ga 242663
it's not possible to stamp out these fires, they're always burning

Cooking of any recipes that require a campfire is possible at any burning bonfire.
Cooking at stamped out bonfire is not possible.

attack an enemy (e.g. bare handed) while having buff  'Bonfire's Blessing'
observe PROC at about 30% hitting with (player level)*10 damage with hidden cooldown of  45 seconds
.event stop 1
observe PROC beeing disabled for this buff, but buff / aura still showing
relog
observe buff / aura 'Bonfire's Blessing' beeing gone

- [ ] This pull request can be tested by following the reproduction steps provided in the linked issue
- [ ] This pull request requires further testing. Provide steps to test your changes. If it requires any specific setup e.g multiple players please specify it as well.

1.
2.
3.

## Known Issues and TODO List:

- [ ]
- [ ]

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

File: src/server/scripts/Events/midsummer.cpp [hunk 1]
```diff
@@ -49,6 +68,319 @@ class go_midsummer_bonfire : public GameObjectScript
...
+    { { 571, 2817, TEAM_ALLIANCE }, &BonfireStampedOutState[72] },
+    { { 571, 2817, TEAM_HORDE    }, &BonfireStampedOutState[73] },
+    { { 571, 3537, TEAM_ALLIANCE }, &BonfireStampedOutState[74] },
+    { { 571, 3537, TEAM_HORDE    }, &BonfireStampedOutState[75] },
+    { { 571, 3711, TEAM_ALLIANCE }, &BonfireStampedOutState[76] },
+    { { 571, 3711, TEAM_HORDE    }, &BonfireStampedOutState[77] },
+};
+
+const uint32 GoBonfireAlliance[COUNT_GO_BONFIRE_ALLIANCE] = { 187564, 187914, 187916, 187917, 187919, 187920, 187921, 187922, 187923, 187924, 187925, 187926, 187927, 187928, 187929, 187930, 187931, 187932, 187933, 187934, 187935, 187936, 187937, 187938, 187939, 187940, 187941, 187942, 187943, 187944, 187945, 187946, 194032, 194035, 194036, 194038, 194040, 194044, 194045, 194049 };
+const uint32 GoBonfireHorde[COUNT_GO_BONFIRE_HORDE] = { 187559, 187947, 187948, 187949, 187950, 187951, 187952, 187953, 187954, 187955, 187956, 187957, 187958, 187959, 187960, 187961, 187962, 187963, 187964, 187965, 187966, 187967, 187968, 187969, 187970, 187971, 187972, 187973, 187974, 187975, 194033, 194034, 194037, 194039, 194042, 194043, 194046, 194048 };
```
Comment (Kitzunu) [inline src/server/scripts/Events/midsummer.cpp:163]: uint32 const go

File: src/server/scripts/Events/midsummer.cpp [hunk 2]
```diff
@@ -49,6 +68,319 @@ class go_midsummer_bonfire : public GameObjectScript
...
+const uint32 GoBonfireCity[COUNT_GO_BONFIRE_CITY] = { 181332, 181333, 181334, 181335, 181336, 181337, 188128, 188129, 188352 };
```
Comment (Kitzunu) [inline src/server/scripts/Events/midsummer.cpp:164]: same here

File: src/server/scripts/Events/midsummer.cpp [hunk 3]
```diff
@@ -49,6 +68,319 @@ class go_midsummer_bonfire : public GameObjectScript
...
+};
+
+struct npc_midsummer_bonfire : public ScriptedAI
+{
+    npc_midsummer_bonfire(Creature* creature) : ScriptedAI(creature)
+    {
+        _isStampedOut  = nullptr;
+        _teamId     = TEAM_NEUTRAL;
+        _type       = BONFIRE_TYPE_NONE;
+        _spellFocus = nullptr;
```
Comment (Kitzunu) [inline src/server/scripts/Events/midsummer.cpp:201]: only private members should prefix with underscore

Comment (Kitzunu) [reply src/server/scripts/Events/midsummer.cpp:201]: https://www.azerothcore.org/wiki/cpp-code-standards#naming-of-variables-and-functions

File: src/server/scripts/Events/midsummer.cpp [hunk 4]
```diff
@@ -49,6 +68,319 @@ class go_midsummer_bonfire : public GameObjectScript
...
+            return false;
+
+        auto itr = BonfireStateStore.find(std::make_tuple(me->GetMapId(), me->GetZoneId(), _teamId));
+        if ((itr != BonfireStateStore.end()) && (itr->second))
+        {
+            _isStampedOut = itr->second;
+        }
+        else
+        {
+            LOG_ERROR("scripts.midsummer", "NPC {} (GUID{}) in map {}, zone {} with teamId {} can't locate its entry within BonfireStateStore", me->GetGUID().GetEntry(), me->GetSpawnId(), me->GetMapId(), me->GetZoneId(), _teamId);
```
Comment (Kitzunu) [inline src/server/scripts/Events/midsummer.cpp:337]: is scripts.midsummer added in the conf file?

File: src/server/scripts/Events/midsummer.cpp (hunk 3 above)
Comment (Kitzunu) [reply src/server/scripts/Events/midsummer.cpp:201]: My bad, I am just tired.

File: src/server/scripts/Events/midsummer.cpp [hunk 5]
```diff
@@ -172,6 +504,25 @@ struct npc_midsummer_torch_target : public ScriptedAI
...
+class spell_bonfires_blessing : public AuraScript
+{
+    PrepareAuraScript(spell_bonfires_blessing)
+
+    void OnApply(AuraEffect const* /*aurEff*/, AuraEffectHandleModes /*mode*/)
+    {
+        if (!IsHolidayActive(HOLIDAY_FIRE_FESTIVAL))
+        {
+            if (Unit* target = GetTarget())
+                target->RemoveAurasDueToSpell(SPELL_BONFIRES_BLESSING);
```
Comment (Kitzunu) [inline src/server/scripts/Events/midsummer.cpp:516]: You need to validate spells used within spell scripts. See Bool Validate(SpellInfo* spellInfo)

File: src/server/scripts/Events/midsummer.cpp (hunk 4 above)
Comment (sudlud) [reply src/server/scripts/Events/midsummer.cpp:337]: No 👀 maybe I'll just drop the error message for now

File: src/server/scripts/Events/midsummer.cpp (hunk 5 above)
Comment (sudlud) [reply src/server/scripts/Events/midsummer.cpp:516]: Alright, I was unsure if it's necessary if the spell is just being used for removeAura()

File: src/server/scripts/Events/midsummer.cpp (hunk 4 above)
Comment (Kitzunu) [reply src/server/scripts/Events/midsummer.cpp:337]: just add it to the conf file. LOG_ERROR will still send iirc but we still want it in conf

File: src/server/scripts/Events/midsummer.cpp (hunk 5 above)
Comment (Kitzunu) [reply src/server/scripts/Events/midsummer.cpp:516]: always, becuase if it for some reason wouldnt exist the server would crash

File: src/server/scripts/Events/midsummer.cpp [hunk 6]
```diff
@@ -49,6 +69,319 @@ class go_midsummer_bonfire : public GameObjectScript
...
+    }
+
+    void Ignite()
+    {
+        if (_isStampedOut)
+            *_isStampedOut = false;
+
+        if (!_spellFocus)
+        {
+            me->CastSpell(me, SPELL_MIDSUMMER_BONFIRE_BUNNIES_2, true);
```
Comment (Kitzunu) [inline src/server/scripts/Events/midsummer.cpp:221]: Can use DoCastSelf, right?

Comment (Kitzunu) [reply src/server/scripts/Events/midsummer.cpp:221]: all of them

File: src/server/scripts/Events/midsummer.cpp [hunk 7]
```diff
@@ -49,6 +69,319 @@ class go_midsummer_bonfire : public GameObjectScript
...
+                    {
+                        if (*_isStampedOut)
+                        {
+                            p->RemoveAurasDueToSpell(SPELL_BONFIRES_BLESSING);
+                        }
+                        else
+                        {
+                            if (!p->HasAura(SPELL_BONFIRES_BLESSING))
+                                p->CastSpell(p, SPELL_BONFIRES_BLESSING, true);
+                        }
```
Comment (Kitzunu) [inline src/server/scripts/Events/midsummer.cpp:285]: ok my man, at least choose one style of brackets, don't mix them 😅

File: src/server/scripts/Events/midsummer.cpp (hunk 6 above)
Comment (sudlud) [reply src/server/scripts/Events/midsummer.cpp:221]: sure, if that's preferred

File: src/server/scripts/Events/midsummer.cpp (hunk 7 above)
Comment (sudlud) [reply src/server/scripts/Events/midsummer.cpp:285]: I usually do brackets if there's more than one statement, is this already "mixing styles"?

Comment (sudlud) [reply src/server/scripts/Events/midsummer.cpp:285]: also if the else clause has brackets the if should also have them? but whatever you prefer i can update it

**Review by Nyeriah — APPROVED** [review]

**Pull request merged**